File: packages/core/ios/GlobalAttributesHandler.swift
//
// Copyright 2025 Splunk Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import Foundation
import React
import SplunkAgent
import OpenTelemetryApi

/// Handles all global attributes operations.
@objcMembers
@objc(GlobalAttributesHandler)
public class GlobalAttributesHandler: NSObject {

  // MARK: - Setters

  public func setString(_ key: NSString,
                        value: NSString?,
                        resolve: @escaping RCTPromiseResolveBlock,
                        reject: @escaping RCTPromiseRejectBlock) {
    if let v = value as String? {
      SplunkRum.shared.globalAttributes.setString(v, for: key as String)
    } else {
      _ = SplunkRum.shared.globalAttributes.remove(for: key as String)
    }

    resolve(nil)
  }

  public func setBoolean(_ key: NSString,
                         value: NSNumber?,
                         resolve: @escaping RCTPromiseResolveBlock,
                         reject: @escaping RCTPromiseRejectBlock) {
    if let v = value?.boolValue {
      SplunkRum.shared.globalAttributes.setBool(v, for: key as String)
    } else {
      _ = SplunkRum.shared.globalAttributes.remove(for: key as String)
    }

    resolve(nil)
  }

  public func setNumber(_ key: NSString,
                        value: NSNumber?,
                        resolve: @escaping RCTPromiseResolveBlock,
                        reject: @escaping RCTPromiseRejectBlock) {
    if let v = value {
      SplunkRum.shared.globalAttributes.setDouble(v.doubleValue, for: key as String)
    } else {
      _ = SplunkRum.shared.globalAttributes.remove(for: key as String)
    }

    resolve(nil)
  }

  public func setArray(_ key: NSString,
                       value: NSArray?,
                       resolve: @escaping RCTPromiseResolveBlock,
                       reject: @escaping RCTPromiseRejectBlock) {
    if let arr = value as? [Any] {
      let mapped = AttributeConverter.convertArrayToAttributeArray(arr)
      SplunkRum.shared.globalAttributes.setArray(AttributeArray(values: mapped), for: key as String)
    } else {
      _ = SplunkRum.shared.globalAttributes.remove(for: key as String)
    }

    resolve(nil)
  }

  // MARK: - Getters

  public func getValue(_ key: NSString,
                       resolve: @escaping RCTPromiseResolveBlock,
                       reject: @escaping RCTPromiseRejectBlock) {
    let v = SplunkRum.shared.globalAttributes.getValue(for: key as String)
    resolve(AttributeConverter.attributeValueToAny(v))
  }

  public func getString(_ key: NSString,
                        resolve: @escaping RCTPromiseResolveBlock,
                        reject: @escaping RCTPromiseRejectBlock) {
    resolve(SplunkRum.shared.globalAttributes.getString(for: key as String))
  }

  public func getBoolean(_ key: NSString,
                         resolve: @escaping RCTPromiseResolveBlock,
                         reject: @escaping RCTPromiseRejectBlock) {
    resolve(SplunkRum.shared.globalAttributes.getBool(for: key as String))
  }

  public func getNumber(_ key: NSString,
                        resolve: @escaping RCTPromiseResolveBlock,
                        reject: @escaping RCTPromiseRejectBlock) {
    if let d = SplunkRum.shared.globalAttributes.getDouble(for: key as String) {
      resolve(NSNumber(value: d))
    } else if let i = SplunkRum.shared.globalAttributes.getInt(for: key as String) {
      resolve(NSNumber(value: i))
    } else {
      resolve(nil)
    }
  }

  public func getArray(_ key: NSString,
                       resolve: @escaping RCTPromiseResolveBlock,
                       reject: @escaping RCTPromiseRejectBlock) {
    if let arr = SplunkRum.shared.globalAttributes.getArray(for: key as String) {
      resolve(arr.values.map { AttributeConverter.attributeValueToAny($0) as Any })
    } else {
      resolve(nil)
    }
  }

  // MARK: - Bulk Operations

  public func setAll(_ map: NSDictionary,
                     resolve: @escaping RCTPromiseResolveBlock,
                     reject: @escaping RCTPromiseRejectBlock) {
    let attrs = AttributeConverter.buildAttributes(from: map)
    let count = SplunkRum.shared.globalAttributes.addDictionary(attrs.getAll())

    resolve(NSNumber(value: count))
  }

  public func setAllInNameSpace(_ nameSpace: NSString,
                                map: NSDictionary,
                                resolve: @escaping RCTPromiseResolveBlock,
                                reject: @escaping RCTPromiseRejectBlock) {
    let attrs = AttributeConverter.buildAttributes(from: map)
    let count = SplunkRum.shared.globalAttributes.addDictionary(attrs.getAll(), intoNamespace: nameSpace as String)

    resolve(NSNumber(value: count))
  }

  public func remove(_ key: NSString,
                     resolve: @escaping RCTPromiseResolveBlock,
                     reject: @escaping RCTPromiseRejectBlock) {
    let removed = SplunkRum.shared.globalAttributes.remove(for: key as String)

    resolve(AttributeConverter.attributeValueToAny(removed))
  }

  public func removeAll(_ resolve: @escaping RCTPromiseResolveBlock,
                        reject: @escaping RCTPromiseRejectBlock) {
    SplunkRum.shared.globalAttributes.removeAll()
    resolve(nil)
  }

  public func contains(_ key: NSString,
                       resolve: @escaping RCTPromiseResolveBlock,
                       reject: @escaping RCTPromiseRejectBlock) {
    resolve(NSNumber(value: SplunkRum.shared.globalAttributes.contains(key: key as String)))
  }

  public func getAll(_ resolve: @escaping RCTPromiseResolveBlock,
                     reject: @escaping RCTPromiseRejectBlock) {
    resolve(SplunkRum.shared.globalAttributes.all)
  }

  public func keys(_ resolve: @escaping RCTPromiseResolveBlock,
                   reject: @escaping RCTPromiseRejectBlock) {
    resolve(SplunkRum.shared.globalAttributes.getAllKeys())
  }

  public func values(_ resolve: @escaping RCTPromiseResolveBlock,
                     reject: @escaping RCTPromiseRejectBlock) {
    let vals = SplunkRum.shared.globalAttributes.getAllValues().map { AttributeConverter.attributeValueToAny($0) as Any }
    resolve(vals)
  }

  public func size(_ resolve: @escaping RCTPromiseResolveBlock,
                   reject: @escaping RCTPromiseRejectBlock) {
    resolve(NSNumber(value: SplunkRum.shared.globalAttributes.count()))
  }
}

File: packages/core/android/src/main/kotlin/com/splunk/otel/reactnative/GlobalAttributesHandler.kt
/*
 * Copyright 2025 Splunk Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.splunk.otel.reactnative

import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableNativeArray
import com.facebook.react.bridge.WritableNativeMap
import com.splunk.rum.integration.agent.api.SplunkRum

/**
 * Handles all global attributes operations.
 */
class GlobalAttributesHandler {

  // MARK: - Setters

  fun setString(key: String, value: String?, promise: Promise) {
    try {
      val ga = SplunkRum.instance.globalAttributes
      if (value == null) ga.remove(key) else ga[key] = value
      promise.resolve(null)
    } catch (t: Throwable) {
      promise.reject("E_GA_SET_STRING", t)
    }
  }

  fun setBoolean(key: String, value: Boolean?, promise: Promise) {
    try {
      val ga = SplunkRum.instance.globalAttributes
      if (value == null) ga.remove(key) else ga[key] = value
      promise.resolve(null)
    } catch (t: Throwable) {
      promise.reject("E_GA_SET_BOOLEAN", t)
    }
  }

  fun setNumber(key: String, value: Double?, promise: Promise) {
    try {
      val ga = SplunkRum.instance.globalAttributes
      if (value == null) {
        ga.remove(key)
      } else {
        val longLike = value.toLong()
        if (value == longLike.toDouble()) ga[key] = longLike else ga[key] = value
      }

      promise.resolve(null)
    } catch (t: Throwable) {
      promise.reject("E_GA_SET_NUMBER", t)
    }
  }

  fun setArray(key: String, value: ReadableArray?, promise: Promise) {
    try {
      val ga = SplunkRum.instance.globalAttributes
      if (value == null) {
        ga.remove(key)
      } else {
        AttributeConverter.setArrayToMutableAttributes(ga, key, value)
      }

      promise.resolve(null)
    } catch (t: Throwable) {
      promise.reject("E_GA_SET_ARRAY", t)
    }
  }

  // MARK: - Getters

  fun getValue(key: String, promise: Promise) {
    try {
      val v = getAnyValue(key)
      promise.resolve(v)
    } catch (t: Throwable) {
      promise.reject("E_GA_GET_VALUE", t)
    }
  }

  fun getString(key: String, promise: Promise) {
    try {
      val ga = SplunkRum.instance.globalAttributes
      val v = ga.get<String>(key)

      promise.resolve(v)
    } catch (t: Throwable) {
      promise.reject("E_GA_GET_STRING", t)
    }
  }

  fun getBoolean(key: String, promise: Promise) {
    try {
      val ga = SplunkRum.instance.globalAttributes
      val v = ga.get<Boolean>(key)

      promise.resolve(v)
    } catch (t: Throwable) {
      promise.reject("E_GA_GET_BOOLEAN", t)
    }
  }

  fun getNumber(key: String, promise: Promise) {
    try {
      val ga = SplunkRum.instance.globalAttributes
      val l = ga.get<Long>(key)

      if (l != null) {
        promise.resolve(l.toDouble())
      } else {
        val d = ga.get<Double>(key)
        promise.resolve(d)
      }
    } catch (t: Throwable) {
      promise.reject("E_GA_GET_NUMBER", t)
    }
  }

  fun getArray(key: String, promise: Promise) {
    try {
      val array = getArrayValue(key)
      promise.resolve(array)
    } catch (t: Throwable) {
      promise.reject("E_GA_GET_ARRAY", t)
    }
  }

  // MARK: - Bulk Operations

  fun setAll(map: ReadableMap, promise: Promise) {
    try {
      val ga = SplunkRum.instance.globalAttributes
      var count = 0

      map.entryIterator.forEachRemaining { entry ->
        AttributeConverter.putDynamicToMutableAttributes(ga, entry.key, entry.value)
        count++
      }

      promise.resolve(count)
    } catch (t: Throwable) {
      promise.reject("E_GA_SET_ALL", t)
    }
  }

  fun setAllInNameSpace(nameSpace: String, map: ReadableMap, promise: Promise) {
    try {
      val ga = SplunkRum.instance.globalAttributes
      var count = 0

      map.entryIterator.forEachRemaining { entry ->
        AttributeConverter.putDynamicToMutableAttributes(ga, "$nameSpace.${entry.key}", entry.value)
        count++
      }

      promise.resolve(count)
    } catch (t: Throwable) {
      promise.reject("E_GA_SET_ALL_NS", t)
    }
  }

  fun remove(key: String, promise: Promise) {
    try {
      val prev = getAnyValue(key)
      SplunkRum.instance.globalAttributes.remove(key)

      promise.resolve(prev)
    } catch (t: Throwable) {
      promise.reject("E_GA_REMOVE", t)
    }
  }

  fun removeAll(promise: Promise) {
    try {
      SplunkRum.instance.globalAttributes.removeAll()
      promise.resolve(null)
    } catch (t: Throwable) {
      promise.reject("E_GA_REMOVE_ALL", t)
    }
  }

  fun contains(key: String, promise: Promise) {
    try {
      val contains = SplunkRum.instance.globalAttributes.contains(key)
      promise.resolve(contains)
    } catch (t: Throwable) {
      promise.reject("E_GA_CONTAINS", t)
    }
  }

  fun getAll(promise: Promise) {
    try {
      val attrs = SplunkRum.instance.globalAttributes.getAll()
      val map = WritableNativeMap()

      attrs.asMap().forEach { (k, v) ->
        when (v) {
          is String -> map.putString(k.key, v)
          is Boolean -> map.putBoolean(k.key, v)
          is Long -> map.putDouble(k.key, v.toDouble())
          is Double -> map.putDouble(k.key, v)
          is List<*> -> map.putArray(k.key, AttributeConverter.listToWritableArray(v))
        }
      }

      promise.resolve(map)
    } catch (t: Throwable) {
      promise.reject("E_GA_GET_ALL", t)
    }
  }

  fun keys(promise: Promise) {
    try {
      val attrs = SplunkRum.instance.globalAttributes.getAll()
      val array = WritableNativeArray()

      attrs.asMap().keys.forEach { array.pushString(it.key) }

      promise.resolve(array)
    } catch (t: Throwable) {
      promise.reject("E_GA_KEYS", t)
    }
  }

  fun values(promise: Promise) {
    try {
      val attrs = SplunkRum.instance.globalAttributes.getAll()
      val array = WritableNativeArray()

      attrs.asMap().values.forEach { v ->
        when (v) {
          is String -> array.pushString(v)
          is Boolean -> array.pushBoolean(v)
          is Long -> array.pushDouble(v.toDouble())
          is Double -> array.pushDouble(v)
          is List<*> -> array.pushArray(AttributeConverter.listToWritableArray(v))
        }
      }

      promise.resolve(array)
    } catch (t: Throwable) {
      promise.reject("E_GA_VALUES", t)
    }
  }

  fun size(promise: Promise) {
    try {
      val size = SplunkRum.instance.globalAttributes.size()
      promise.resolve(size)
    } catch (t: Throwable) {
      promise.reject("E_GA_SIZE", t)
    }
  }

  // MARK: - Private Helpers

  private fun getAnyValue(key: String): Any? {
    val ga = SplunkRum.instance.globalAttributes
    ga.get<String>(key)?.let { return it }
    ga.get<Boolean>(key)?.let { return it }
    ga.get<Long>(key)?.let { return it.toDouble() }
    ga.get<Double>(key)?.let { return it }

    val entry = SplunkRum.instance.globalAttributes.getAll().asMap().entries.firstOrNull { it.key.key == key }
    return when (val v = entry?.value) {
      is List<*> -> AttributeConverter.listToWritableArray(v)
      else -> null
    }
  }

  private fun getArrayValue(key: String): WritableNativeArray? {
    val entry = SplunkRum.instance.globalAttributes.getAll().asMap().entries.firstOrNull { it.key.key == key }
    
    return when (val v = entry?.value) {
      is List<*> -> AttributeConverter.listToWritableArray(v)
      else -> null
    }
  }
}
